### DISCUSSION TITLE
error : Install  C++ interface

### DISCUSSION THREAD

**Discussion by BZBJ224** [deepmodeling/deepmd-kit#332] [Q&A]
Hello,When I installed the DeePMD-kit's C++ interface  the following error occurred:

[ 97%] Linking CXX executable dp_ipi
../op/libdeepmd_op.so：对‘tensorflow::kernel_factory::OpKernelRegistrar::InitInternal(tensorflow::KernelDef const*, absl::string_view, std::unique_ptr<tensorflow::kernel_factory::OpKernelFactory, std::default_delete<tensorflow::kernel_factory::OpKernelFactory> >)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::Status::Status(tensorflow::error::Code, absl::string_view)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::register_op::OpDefBuilderReceiver::OpDefBuilderReceiver(tensorflow::register_op::OpDefBuilderWrapper<true> const&)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, float*)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::OpKernelContext::input(int)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, int*)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, std::vector<int, std::allocator<int> >*)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::TensorShapeBase<tensorflow::TensorShape>::AddDim(long long)’未定义的引用
collect2: error: ld returned 1 exit status
ipi/CMakeFiles/dp_ipi.dir/build.make:102: recipe for target 'ipi/dp_ipi' failed
make[2]: *** [ipi/dp_ipi] Error 1
CMakeFiles/Makefile2:339: recipe for target 'ipi/CMakeFiles/dp_ipi.dir/all' failed
make[1]: *** [ipi/CMakeFiles/dp_ipi.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2

environment:
ubuntu18.04  
gcc--7.5.0 bazel--3.7.2 cuda--11.1  

So what is the cause of this error, and how to solve it?

**Comment by dingye18** [comment]
Did you add the tensorflow library path into LD_LIBRARY_PATH before?

**Reply by BZBJ224** [reply]
Yes, export lib in file ~/.bashrc.

**Reply by dingye18** [reply]
Can you add the `SET(CMAKE_VERBOSE_MAKEFILE ON)` at the head of CMakeLists.txt in source directory. And paste the output of `make` command into this issue? It might print out more useful message during compiling.

**Reply by BZBJ224** [reply]
The output content after add the SET(CMAKE_VERBOSE_MAKEFILE ON) is as follows:

[ 97%] Linking CXX executable dp_ipi
/usr/bin/c++   -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG   -Wl,--no-as-needed -Wl,-rpath,'$ORIGIN'/../lib -Wl,-z,defs CMakeFiles/dp_ipi.dir/driver.cc.o  -o dp_ipi -Wl,-rpath,/home/ycx/下载/deepmd-kit/source/build/ipi:/home/ycx/下载/deepmd-kit/source/build/op:/home/ycx/下载/deepmd-kit/source/build/lib:/home/ycx/下载/deepmd-kit/source/11:/home/ycx/下载/deepmd-kit/source/build/op/cuda: libdeepmd_ipi.so ../op/libdeepmd_op.so ../lib/libdeepmd.so ../../11/libtensorflow_cc.so ../../11/libtensorflow_framework.so.2 ../op/cuda/libdeepmd_op_cuda.so /usr/local/cuda/lib64/libcudart_static.a -lpthread -ldl /usr/lib/x86_64-linux-gnu/librt.so 
../op/libdeepmd_op.so：对‘tensorflow::kernel_factory::OpKernelRegistrar::InitInternal(tensorflow::KernelDef const*, absl::string_view, std::unique_ptr<tensorflow::kernel_factory::OpKernelFactory, std::default_delete<tensorflow::kernel_factory::OpKernelFactory> >)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::Status::Status(tensorflow::error::Code, absl::string_view)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::register_op::OpDefBuilderReceiver::OpDefBuilderReceiver(tensorflow::register_op::OpDefBuilderWrapper<true> const&)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, float*)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::OpKernelContext::input(int)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, int*)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, std::vector<int, std::allocator<int> >*)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::TensorShapeBase<tensorflow::TensorShape>::AddDim(long long)’未定义的引用
collect2: error: ld returned 1 exit status
ipi/CMakeFiles/dp_ipi.dir/build.make:105: recipe for target 'ipi/dp_ipi' failed
make[2]: *** [ipi/dp_ipi] Error 1
CMakeFiles/Makefile2:342: recipe for target 'ipi/CMakeFiles/dp_ipi.dir/all' failed
make[1]: *** [ipi/CMakeFiles/dp_ipi.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2

**Reply by dingye18** [reply]
Check the output of `ldd ./op/libdeepmd_op.so` in your build directory? If it is linked with tensorflow library?

**Reply by dingye18** [reply]
If not, please add the `target_link_libraries(${LIB_DEEPMD_OP} ${TensorFlowFramework_LIBRARY})` into the CMakeLists.txt locate in `source/op` directory. Refer to this figure for which line you write in CMakeLists.txt.

![image](https://user-images.githubusercontent.com/13417572/105180009-fcbc7b00-5b64-11eb-8908-bf1bc5db5aae.png)

**Reply by BZBJ224** [reply]
After adding this condition in source/op/CMakeLists.txt, It is linked with tensorflow library, and added 'libtensorflow_framework.so.2 => /home/ycx/下载/deepmd-kit/source/11/libtensorflow_framework.so.2 (0x00007f93b96cf000)' in the output.  But the error still exists.

[ 97%] Linking CXX executable dp_ipi
cd /home/ycx/下载/deepmd-kit/source/build/ipi && /usr/bin/cmake -E cmake_link_script CMakeFiles/dp_ipi.dir/link.txt --verbose=1
/usr/bin/c++   -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG   -Wl,--no-as-needed -Wl,-rpath,'$ORIGIN'/../lib -Wl,-z,defs CMakeFiles/dp_ipi.dir/driver.cc.o  -o dp_ipi -Wl,-rpath,/home/ycx/下载/deepmd-kit/source/build/ipi:/home/ycx/下载/deepmd-kit/source/build/op:/home/ycx/下载/deepmd-kit/source/build/lib:/home/ycx/下载/deepmd-kit/source/11:/home/ycx/下载/deepmd-kit/source/build/op/cuda: libdeepmd_ipi.so ../op/libdeepmd_op.so ../lib/libdeepmd.so ../../11/libtensorflow_cc.so ../../11/libtensorflow_framework.so.2 ../op/cuda/libdeepmd_op_cuda.so /usr/local/cuda/lib64/libcudart_static.a -lpthread -ldl /usr/lib/x86_64-linux-gnu/librt.so 
../op/libdeepmd_op.so：对‘tensorflow::kernel_factory::OpKernelRegistrar::InitInternal(tensorflow::KernelDef const*, absl::string_view, std::unique_ptr<tensorflow::kernel_factory::OpKernelFactory, std::default_delete<tensorflow::kernel_factory::OpKernelFactory> >)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::Status::Status(tensorflow::error::Code, absl::string_view)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::register_op::OpDefBuilderReceiver::OpDefBuilderReceiver(tensorflow::register_op::OpDefBuilderWrapper<true> const&)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, float*)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::OpKernelContext::input(int)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, int*)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, std::vector<int, std::allocator<int> >*)’未定义的引用
../op/libdeepmd_op.so：对‘tensorflow::TensorShapeBase<tensorflow::TensorShape>::AddDim(long long)’未定义的引用
collect2: error: ld returned 1 exit status
ipi/CMakeFiles/dp_ipi.dir/build.make:105: recipe for target 'ipi/dp_ipi' failed
make[2]: *** [ipi/dp_ipi] Error 1
make[2]: 离开目录“/home/ycx/下载/deepmd-kit/source/build”
CMakeFiles/Makefile2:342: recipe for target 'ipi/CMakeFiles/dp_ipi.dir/all' failed
make[1]: *** [ipi/CMakeFiles/dp_ipi.dir/all] Error 2
make[1]: 离开目录“/home/ycx/下载/deepmd-kit/source/build”
Makefile:132: recipe for target 'all' failed
make: *** [all] Error 2

**Reply by dingye18** [reply]
What's the output when it compiled libdeepmd_op.so? Can you paste the full output of compiling into this discussion? Besides, it looks like you used Chinese Language as default language of Ubuntu, how about change it into English?

**Reply by BZBJ224** [reply]
The output is  
	'linux-vdso.so.1 (0x00007fff9a3db000)
	libdeepmd_op_cuda.so => /home/ycx/Downloads/deepmd-kit/source/build/op/cuda/libdeepmd_op_cuda.so (0x00007f957cecc000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f957ccad000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f957caa9000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f957c8a1000)
	libtensorflow_framework.so.2 => /home/ycx/Downloads/deepmd-kit/source/11/libtensorflow_framework.so.2 (0x00007f957a90d000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f957a584000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f957a1e6000)
	libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f9579fb7000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f9579d9f000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f95799ae000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f957d5b4000)'

And I change language to English, the error as follows:
[100%] Linking CXX executable dp_ipi
../op/libdeepmd_op.so: undefined reference to `tensorflow::kernel_factory::OpKernelRegistrar::InitInternal(tensorflow::KernelDef const*, absl::string_view, std::unique_ptr<tensorflow::kernel_factory::OpKernelFactory, std::default_delete<tensorflow::kernel_factory::OpKernelFactory> >)'
../op/libdeepmd_op.so: undefined reference to `tensorflow::Status::Status(tensorflow::error::Code, absl::string_view)'
../op/libdeepmd_op.so: undefined reference to `tensorflow::register_op::OpDefBuilderReceiver::OpDefBuilderReceiver(tensorflow::register_op::OpDefBuilderWrapper<true> const&)'
../op/libdeepmd_op.so: undefined reference to `tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, float*)'
../op/libdeepmd_op.so: undefined reference to `tensorflow::OpKernelContext::input(int)'
../op/libdeepmd_op.so: undefined reference to `tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, int*)'
../op/libdeepmd_op.so: undefined reference to `tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, std::vector<int, std::allocator<int> >*)'
../op/libdeepmd_op.so: undefined reference to `tensorflow::TensorShapeBase<tensorflow::TensorShape>::AddDim(long long)'
collect2: error: ld returned 1 exit status
ipi/CMakeFiles/dp_ipi.dir/build.make:102: recipe for target 'ipi/dp_ipi' failed
make[2]: *** [ipi/dp_ipi] Error 1
CMakeFiles/Makefile2:339: recipe for target 'ipi/CMakeFiles/dp_ipi.dir/all' failed
make[1]: *** [ipi/CMakeFiles/dp_ipi.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2

**Reply by dingye18** [reply]
I mean the compiling output of when you use `make` command.

**Reply by BZBJ224** [reply]
[ 62%] Linking CXX shared library libdeepmd_op.so
cd /home/ycx/Downloads/deepmd-kit/source/build/op && /usr/bin/cmake -E cmake_link_script CMakeFiles/deepmd_op.dir/link.txt --verbose=1
/usr/bin/c++ -fPIC  -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG  -Wl,--no-as-needed -shared -Wl,-soname,libdeepmd_op.so -o libdeepmd_op.so CMakeFiles/deepmd_op.dir/prod_force.cc.o CMakeFiles/deepmd_op.dir/prod_virial.cc.o CMakeFiles/deepmd_op.dir/descrpt.cc.o CMakeFiles/deepmd_op.dir/descrpt_se_a_multi_device.cc.o CMakeFiles/deepmd_op.dir/descrpt_se_r_multi_device.cc.o CMakeFiles/deepmd_op.dir/tab_inter.cc.o CMakeFiles/deepmd_op.dir/prod_force_se_a_multi_device.cc.o CMakeFiles/deepmd_op.dir/prod_virial_se_a_multi_device.cc.o CMakeFiles/deepmd_op.dir/prod_force_se_r_multi_device.cc.o CMakeFiles/deepmd_op.dir/prod_virial_se_r_multi_device.cc.o CMakeFiles/deepmd_op.dir/soft_min.cc.o CMakeFiles/deepmd_op.dir/soft_min_force.cc.o CMakeFiles/deepmd_op.dir/soft_min_virial.cc.o CMakeFiles/deepmd_op.dir/gelu_multi_device.cc.o -Wl,-rpath,/home/ycx/Downloads/deepmd-kit/source/build/op/cuda:/home/ycx/Downloads/deepmd-kit/source/11: cuda/libdeepmd_op_cuda.so /usr/local/cuda/lib64/libcudart_static.a -lpthread -ldl /usr/lib/x86_64-linux-gnu/librt.so ../../11/libtensorflow_framework.so 
/usr/bin/cmake -E __run_co_compile --lwyu=libdeepmd_op.so
make[2]: Leaving directory '/home/ycx/Downloads/deepmd-kit/source/build'
[ 62%] Built target deepmd_op
make -f lib/CMakeFiles/deepmd.dir/build.make lib/CMakeFiles/deepmd.dir/depend
make[2]: Entering directory '/home/ycx/Downloads/deepmd-kit/source/build'
cd /home/ycx/Downloads/deepmd-kit/source/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/ycx/Downloads/deepmd-kit/source /home/ycx/Downloads/deepmd-kit/source/lib /home/ycx/Downloads/deepmd-kit/source/build /home/ycx/Downloads/deepmd-kit/source/build/lib /home/ycx/Downloads/deepmd-kit/source/build/lib/CMakeFiles/deepmd.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/ycx/Downloads/deepmd-kit/source/build'
make -f lib/CMakeFiles/deepmd.dir/build.make lib/CMakeFiles/deepmd.dir/build
make[2]: Entering directory '/home/ycx/Downloads/deepmd-kit/source/build'
[ 64%] Building CXX object lib/CMakeFiles/deepmd.dir/src/DataModifier.cc.o
cd /home/ycx/Downloads/deepmd-kit/source/build/lib && /usr/bin/c++  -DHIGH_PREC -Ddeepmd_EXPORTS -I/home/ycx/Downloads/deepmd-kit/source/lib/include -I/home/ycx/Downloads/deepmd-kit/source/3rdparty -I/home/ycx/Downloads/deepmd-kit/source/11 -I/home/ycx/Downloads/deepmd-kit/source/build/lib -I/usr/local/cuda/include  -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG -fPIC   -D GOOGLE_CUDA -o CMakeFiles/deepmd.dir/src/DataModifier.cc.o -c /home/ycx/Downloads/deepmd-kit/source/lib/src/DataModifier.cc
[ 67%] Building CXX object lib/CMakeFiles/deepmd.dir/src/DeepTensor.cc.o
cd /home/ycx/Downloads/deepmd-kit/source/build/lib && /usr/bin/c++  -DHIGH_PREC -Ddeepmd_EXPORTS -I/home/ycx/Downloads/deepmd-kit/source/lib/include -I/home/ycx/Downloads/deepmd-kit/source/3rdparty -I/home/ycx/Downloads/deepmd-kit/source/11 -I/home/ycx/Downloads/deepmd-kit/source/build/lib -I/usr/local/cuda/include  -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG -fPIC   -D GOOGLE_CUDA -o CMakeFiles/deepmd.dir/src/DeepTensor.cc.o -c /home/ycx/Downloads/deepmd-kit/source/lib/src/DeepTensor.cc
[ 70%] Building CXX object lib/CMakeFiles/deepmd.dir/src/NNPAtomMap.cc.o
cd /home/ycx/Downloads/deepmd-kit/source/build/lib && /usr/bin/c++  -DHIGH_PREC -Ddeepmd_EXPORTS -I/home/ycx/Downloads/deepmd-kit/source/lib/include -I/home/ycx/Downloads/deepmd-kit/source/3rdparty -I/home/ycx/Downloads/deepmd-kit/source/11 -I/home/ycx/Downloads/deepmd-kit/source/build/lib -I/usr/local/cuda/include  -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG -fPIC   -D GOOGLE_CUDA -o CMakeFiles/deepmd.dir/src/NNPAtomMap.cc.o -c /home/ycx/Downloads/deepmd-kit/source/lib/src/NNPAtomMap.cc
[ 72%] Building CXX object lib/CMakeFiles/deepmd.dir/src/NNPInter.cc.o
cd /home/ycx/Downloads/deepmd-kit/source/build/lib && /usr/bin/c++  -DHIGH_PREC -Ddeepmd_EXPORTS -I/home/ycx/Downloads/deepmd-kit/source/lib/include -I/home/ycx/Downloads/deepmd-kit/source/3rdparty -I/home/ycx/Downloads/deepmd-kit/source/11 -I/home/ycx/Downloads/deepmd-kit/source/build/lib -I/usr/local/cuda/include  -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG -fPIC   -D GOOGLE_CUDA -o CMakeFiles/deepmd.dir/src/NNPInter.cc.o -c /home/ycx/Downloads/deepmd-kit/source/lib/src/NNPInter.cc
[ 75%] Building CXX object lib/CMakeFiles/deepmd.dir/src/common.cc.o
cd /home/ycx/Downloads/deepmd-kit/source/build/lib && /usr/bin/c++  -DHIGH_PREC -Ddeepmd_EXPORTS -I/home/ycx/Downloads/deepmd-kit/source/lib/include -I/home/ycx/Downloads/deepmd-kit/source/3rdparty -I/home/ycx/Downloads/deepmd-kit/source/11 -I/home/ycx/Downloads/deepmd-kit/source/build/lib -I/usr/local/cuda/include  -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG -fPIC   -D GOOGLE_CUDA -o CMakeFiles/deepmd.dir/src/common.cc.o -c /home/ycx/Downloads/deepmd-kit/source/lib/src/common.cc
[ 78%] Building CXX object lib/CMakeFiles/deepmd.dir/src/NeighborList.cpp.o
cd /home/ycx/Downloads/deepmd-kit/source/build/lib && /usr/bin/c++  -DHIGH_PREC -Ddeepmd_EXPORTS -I/home/ycx/Downloads/deepmd-kit/source/lib/include -I/home/ycx/Downloads/deepmd-kit/source/3rdparty -I/home/ycx/Downloads/deepmd-kit/source/11 -I/home/ycx/Downloads/deepmd-kit/source/build/lib -I/usr/local/cuda/include  -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG -fPIC   -D GOOGLE_CUDA -o CMakeFiles/deepmd.dir/src/NeighborList.cpp.o -c /home/ycx/Downloads/deepmd-kit/source/lib/src/NeighborList.cpp
[ 81%] Building CXX object lib/CMakeFiles/deepmd.dir/src/SimulationRegion.cpp.o
cd /home/ycx/Downloads/deepmd-kit/source/build/lib && /usr/bin/c++  -DHIGH_PREC -Ddeepmd_EXPORTS -I/home/ycx/Downloads/deepmd-kit/source/lib/include -I/home/ycx/Downloads/deepmd-kit/source/3rdparty -I/home/ycx/Downloads/deepmd-kit/source/11 -I/home/ycx/Downloads/deepmd-kit/source/build/lib -I/usr/local/cuda/include  -std=c++11 -Wno-ignored-attributes -fopenmp -O3 -DNDEBUG -fPIC   -D GOOGLE_CUDA -o CMakeFiles/deepmd.dir/src/SimulationRegion.cpp.o -c [/home/ycx/Downloads/deepmd-kit/source/lib/src/SimulationRegion.cpp[]
